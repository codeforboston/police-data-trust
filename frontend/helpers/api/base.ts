import axiosModule, { AxiosRequestConfig } from "axios"
import { baseURL } from "./config"

export type AccessToken = string

export interface AuthenticatedRequest {
  accessToken: AccessToken
}

const axios = axiosModule.create({
  baseURL,
  timeout: 5000
})

export function request({
  accessToken,
  ...config
}: AxiosRequestConfig & { accessToken?: AccessToken }) {
  let { headers, ...rest } = config
  if (accessToken) {
    headers = {
      Authorization: `Bearer ${accessToken}`,
      ...headers
    }
  }

  return axios({
    headers,
    ...rest
  }).then((response) => response.data)
}
