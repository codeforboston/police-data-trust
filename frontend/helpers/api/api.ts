import axiosModule, { AxiosRequestConfig } from "axios"

export type AccessToken = string

export interface User {
  active: boolean
  role: string
  email: string
  emailConfirmedAt?: string
  firstName?: string
  lastName?: string
  phoneNumber?: string
}

export interface NewUser {
  email: string
  password: string
  firstName?: string
  lastName?: string
  phoneNumber?: string
}

export interface LoginCredentials {
  email: string
  password: string
}

export interface ForgotPassword {
  email: string
}

export interface ResetPasswordRequest extends AuthenticatedRequest {
  accessToken: string
  password: string
}

export interface ResetPasswordResponse {
  message: string
}

export interface Source {
  name?: string
  id?: number
  url?: string
  contact_email?: string
}

export interface Perpetrator {
  first_name?: string
  last_name?: string
}

export enum Rank {
  TECHNICIAN = "Technician",
  OFFICER = "Officer",
  DETECTIVE = "Detective",
  CORPORAL = "Corporal",
  SERGEANT = "Sergeant",
  LIEUTENANT = "Lieutenant",
  CAPTAIN = "Captain",
  DEPUTY = "Deputy",
  CHIEF = "Chief",
  COMMISSIONER = "Commissioner"
}

export interface Officer {
  id?: number
  first_name?: string
  last_name?: string
  race?: string
  ethnicity?: string
  gender?: string
  rank?: Rank
  star?: string
  date_of_birth?: Date
}

export interface UseOfForce {
  item?: string
}
export interface Incident {
  id: number
  source?: Source
  source_id?: number
  location?: string
  locationLonLat?: [number, number] //TODO: Backend data does not return locationLonLat attribute. Remove this and refactor frontend
  latitude?: number
  longitude?: number
  time_of_incident?: string
  department?: string
  perpetrators: Perpetrator[]
  description?: string
  use_of_force?: UseOfForce[]
}

interface AuthenticatedRequest {
  accessToken: AccessToken
}

export type RegisterRequest = NewUser
export type LoginRequest = LoginCredentials
export type WhoamiRequest = AuthenticatedRequest
export interface IncidentSearchRequest extends AuthenticatedRequest {
  description?: string
  dateStart?: string
  dateEnd?: string
  location?: string
  source?: string
  page?: number
  perPage?: number
}

export type IncidentSearchResponse = {
  results: Incident[]
  page: number
  totalPages: number
  totalResults: number
}

export const baseURL = process.env.NEXT_PUBLIC_API_BASE_URL || "http://localhost:5000/api/v1"

const axios = axiosModule.create({
  baseURL,
  timeout: 5000
})

export function login(data: LoginRequest): Promise<AccessToken> {
  return request({
    url: "/auth/login",
    method: "POST",
    data
  }).then(({ access_token }) => access_token)
}

export function register(data: RegisterRequest): Promise<AccessToken> {
  return request({
    url: "/auth/register",
    method: "POST",
    data
  }).then(({ access_token }) => access_token)
}

export function forgotPassowrd(data: ForgotPassword): Promise<void> {
  return request({
    url: "/auth/forgotPassword",
    method: "POST",
    data
  })
}

export function resetPassword(req: ResetPasswordRequest): Promise<ResetPasswordResponse> {
  const { accessToken } = req

  return request({
    url: `/auth/resetPassword`,
    method: "POST",
    data: { password: req.password },
    accessToken
  })
}

export function whoami({ accessToken }: WhoamiRequest): Promise<User> {
  return request({
    url: "/auth/whoami",
    method: "GET",
    accessToken
  }).then(({ active, email, email_confirmed_at, first_name, last_name, phone_number, role }) => ({
    active,
    email,
    emailConfirmedAt: email_confirmed_at,
    firstName: first_name,
    lastName: last_name,
    phoneNumber: phone_number,
    role: role
  }))
}

export function searchIncidents({
  accessToken,
  dateStart,
  dateEnd,
  ...rest
}: IncidentSearchRequest): Promise<IncidentSearchResponse> {
  if (dateStart) dateStart = new Date(dateStart).toISOString().slice(0, -1)
  if (dateEnd) dateEnd = new Date(dateEnd).toISOString().slice(0, -1)

  return request({
    url: "/incidents/search",
    method: "POST",
    accessToken,
    data: { dateStart, dateEnd, ...rest }
  })
}

export async function getIncidentById(id: number, accessToken: string): Promise<Incident> {
  return request({
    url: `/incidents/get/${id}`,
    method: "GET",
    accessToken
  })
}

function request({ accessToken, ...config }: AxiosRequestConfig & { accessToken?: AccessToken }) {
  let { headers, ...rest } = config
  if (accessToken) {
    headers = {
      Authorization: `Bearer ${accessToken}`,
      ...headers
    }
  }

  return axios({
    headers,
    ...rest
  }).then((response) => response.data)
}
