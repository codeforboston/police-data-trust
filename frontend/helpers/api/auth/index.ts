export * from "./types"
export * from "./auth"
