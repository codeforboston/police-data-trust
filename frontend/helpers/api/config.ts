export const baseURL = process.env.NEXT_PUBLIC_API_BASE_URL || "http://localhost:5000/api/v1"
