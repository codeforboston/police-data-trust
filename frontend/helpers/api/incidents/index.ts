export * from "./types"
export * from "./incidents"
