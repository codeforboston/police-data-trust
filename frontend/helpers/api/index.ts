export * from "./base"
export * from "./config"
export * from "./auth"
export * from "./incidents"
export { useMockServiceWorker, apiMode } from "./mocks/browser.setup"
