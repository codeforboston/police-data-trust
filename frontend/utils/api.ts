export type AccessToken = string

export interface Source {
  name?: string
  id?: number
  url?: string
  contact_email?: string
}

export interface Perpetrator {
  first_name?: string
  last_name?: string
}

export interface UseOfForce {
  item?: string
}
export interface Incident {
  id: number
  source?: Source
  source_id?: number
  location?: string
  locationLonLat?: [number, number] //TODO: Backend data does not return locationLonLat attribute. Remove this and refactor frontend
  latitude?: number
  longitude?: number
  time_of_incident?: string
  department?: string
  perpetrators: Perpetrator[]
  description?: string
  use_of_force?: UseOfForce[]
}

interface AuthenticatedRequest {
  accessToken: AccessToken
}

export interface SearchRequest extends AuthenticatedRequest {
  query: string
  location?: string
  source?: string
}

export type SearchResponse = {
  uid: string | number
  title: string
  subtitle: string
  content_type: string
  source: string
  last_updated: string
  description?: string
  tags?: string[]
}

export type PaginatedSearchResponses = {
  error?: string | null
  page: number
  pages: number
  per_page: number
  total: number
  results: SearchResponse[]
}
