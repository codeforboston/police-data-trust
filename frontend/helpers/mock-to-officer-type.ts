import { OfficerRecordType, EmploymentType } from "../models/officer"
import { PerpetratorRecordType } from "../models/perpetrator"
import { Incident, Officer, UseOfForce } from "../helpers/api"
import officers from "../models/mock-data/officer.json"
import { IncidentRecordType } from "../models/incident"

export function getOfficerFromMockData(officerId: number) {
  if (officerId >= 0 && officerId < 100) {
    return mockToOfficerType(officers[officerId])
  }
}

export function getPerpetratorFromMockData(perpetratorId: number) {
  if (perpetratorId >= 0 && perpetratorId < 100) {
    return mockToPerpetratorType(officers[perpetratorId])
  }
}

export function mockToOfficerType(officer: typeof officers[0]): OfficerRecordType {
  function mockToWorkHistoryType(workHistory: typeof officer.workHistory): EmploymentType[] {
    const converted: EmploymentType[] = workHistory.map((item) => {
      return {
        agency: {
          agencyName: item.deptName,
          agencyImage: item.deptImage.replace("./frontend/models/mock-data/dept-images", ""),
          agencyHqAddress: item.deptAddress,
          websiteUrl: "https://www.google.com/search?q=police+department"
        },
        currentlyEmployed: true,
        earliestEmployment: new Date(item.dates.split("-")[0].trim()),
        latestEmployment: new Date(item.dates.split("-")[1].trim())
      }
    })
    return converted
  }

  return {
    recordId: officer.id,
    firstName: officer.firstName,
    lastName: officer.lastName,
    dateOfBirth: new Date(officer.birthDate),
    gender: officer.gender,
    race: officer.race,
    badgeNo: officer.badgeNo,
    status: officer.status,
    department: officer.department,
    ethnicity: officer.ethnicity,
    affiliations: officer.affiliations,
    workHistory: mockToWorkHistoryType(officer.workHistory),
    incidents: mockToIncidentType(officer.incidents)
  }
}

export function mockToPerpetratorType(officer: typeof officers[0]): PerpetratorRecordType {
  return {
    recordId: officer.id,
    firstName: officer.firstName,
    lastName: officer.lastName,
    gender: officer.gender,
    race: officer.race
  }
}

export function mockToIncidentType(data: IncidentRecordType[]): IncidentRecordType[] {
  return data
}
