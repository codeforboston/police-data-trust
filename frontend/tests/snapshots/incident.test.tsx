import { render } from "../test-utils"
import IncidentView from "../../compositions/incident-view"
import { getOfficerFromMockData } from "../../helpers/mock-to-officer-type"
import { Incident } from "../../helpers/api"

describe("Incident Page", () => {
  it("Renders the incident page correctly", () => {
    const sampleIncident: Incident = {
      id: 0,
      time_of_incident: "March 1, 2022",
      longitude: -71.0589,
      latitude: 42.3601,
      description:
        "Officer issued motorist ticket, use of excessive force resulted in civilian injury.",
      perpetrators: [
        { first_name: "George", last_name: "Lopez" },
        { first_name: "Hannah", last_name: "Montana" }
      ]
    }

    const tree = render(<IncidentView {...sampleIncident} />)
    expect(tree).toMatchSnapshot()
  })
})
