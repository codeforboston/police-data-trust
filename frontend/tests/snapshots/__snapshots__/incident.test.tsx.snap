// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Incident Page Renders the incident page correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
      />
      <div
        class="wrapper"
      >
        <div
          class="idAndStop"
        >
          <strong>
            0
          </strong>
        </div>
        <div
          class="data"
        >
          <p
            class="category"
          >
            Date
          </p>
          <p>
            Tue Mar 01 2022
          </p>
        </div>
        <div
          class="data"
        >
          <p
            class="category"
          >
            Location
          </p>
          <p />
        </div>
      </div>
      <div
        class="bodyWrapper"
      >
        <div
          class="map"
          id="map"
        />
        <div
          class="wrapper"
        >
          <div
            class="dataBlock"
          >
            <p
              class="category"
            >
              Summary:
            </p>
            <p
              class="data"
            >
              Officer issued motorist ticket, use of excessive force resulted in civilian injury.
            </p>
          </div>
          <div
            class="dataBlock"
          >
            <p
              class="category"
            >
              Agencies Involved:
            </p>
          </div>
          <div
            class="dataBlock"
          >
            <p
              class="category"
            >
              Officers Involved:
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
    />
    <div
      class="wrapper"
    >
      <div
        class="idAndStop"
      >
        <strong>
          0
        </strong>
      </div>
      <div
        class="data"
      >
        <p
          class="category"
        >
          Date
        </p>
        <p>
          Tue Mar 01 2022
        </p>
      </div>
      <div
        class="data"
      >
        <p
          class="category"
        >
          Location
        </p>
        <p />
      </div>
    </div>
    <div
      class="bodyWrapper"
    >
      <div
        class="map"
        id="map"
      />
      <div
        class="wrapper"
      >
        <div
          class="dataBlock"
        >
          <p
            class="category"
          >
            Summary:
          </p>
          <p
            class="data"
          >
            Officer issued motorist ticket, use of excessive force resulted in civilian injury.
          </p>
        </div>
        <div
          class="dataBlock"
        >
          <p
            class="category"
          >
            Agencies Involved:
          </p>
        </div>
        <div
          class="dataBlock"
        >
          <p
            class="category"
          >
            Officers Involved:
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
